### PR TITLE
[@types/camelize] fixes types for camelize package

### PR DESCRIPTION
**Pull request by lukaskral**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/ljharb/camelize/blob/main/index.js
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

### CODE REVIEW

File: types/camelize/index.d.ts [hunk 1]
```diff
@@ -7,14 +7,14 @@ type CamelCase<S extends string> = S extends `${infer FirstWord}${CamelSeparator
...
+        : T[K] extends {} ? Camelize<T[K]>
+        : T[K];
 };
 
-declare function camelize<T>(obj: T): T extends string ? CamelCase<T>
-    : T extends RegExp ? T
-    : T extends Date ? T
-    : T extends {} ? Camelize<T>
-    : T;
+declare function camelize<T>(obj: T): T extends string ? CamelCase<T> : Camelize<T>;
```
Comment (lukaskral) [inline types/camelize/index.d.ts:18]: top level exported function only checks whether the provided argument is string
https://github.com/ljharb/camelize/blob/main/index.js#L4

File: types/camelize/index.d.ts [hunk 2]
```diff
@@ -7,14 +7,14 @@ type CamelCase<S extends string> = S extends `${infer FirstWord}${CamelSeparator
...
+    [K in keyof T as CamelCase<string & K>]: T[K] extends Date ? T[K]
+        : T[K] extends RegExp ? T[K]
+        : T[K] extends Array<infer I> ? Array<Camelize<I>>
+        : T[K] extends {} ? Camelize<T[K]>
+        : T[K];
 };
 
-declare function camelize<T>(obj: T): T extends string ? CamelCase<T>
-    : T extends RegExp ? T
-    : T extends Date ? T
```
Comment (lukaskral) [inline types/camelize/index.d.ts:16]: check for `RegExp` and `Date` happens in the `walk` function which is also run on nested fields. Moving this condition to the `Camelize` type

File: types/camelize/index.d.ts [hunk 3]
```diff
@@ -7,14 +7,14 @@ type CamelCase<S extends string> = S extends `${infer FirstWord}${CamelSeparator
     ? `${Lowercase<FirstWord>}${Uppercase<SecondChar>}${CamelCase<Remaining>}`
     : Lowercase<S>;
 
-type Camelize<T> = {
-    [K in keyof T as CamelCase<string & K>]: T[K] extends {} ? Camelize<T[K]> : T[K];
+export type Camelize<T> = {
+    [K in keyof T as CamelCase<string & K>]: T[K] extends Date ? T[K]
+        : T[K] extends RegExp ? T[K]
+        : T[K] extends Array<infer I> ? Array<Camelize<I>>
```
Comment (lukaskral) [inline types/camelize/index.d.ts:13]: I've also added support for arrays (these are [handled separately in the original package](https://github.com/ljharb/camelize/blob/main/index.js#L11) as well)

File: types/camelize/index.d.ts [hunk 4]
```diff
@@ -7,14 +7,14 @@ type CamelCase<S extends string> = S extends `${infer FirstWord}${CamelSeparator
...
-type Camelize<T> = {
-    [K in keyof T as CamelCase<string & K>]: T[K] extends {} ? Camelize<T[K]> : T[K];
-};
-
-declare function camelize<T>(obj: T): T extends string ? CamelCase<T>
+type Camelize<T> = T extends Date ? T
     : T extends RegExp ? T
-    : T extends Date ? T
-    : T extends {} ? Camelize<T>
+    : T extends Array<infer I> ? Array<Camelize<I>>
```
Comment (lukaskral) [inline types/camelize/index.d.ts:12]: I've also added support for arrays (these are [handled separately in the original package](https://github.com/ljharb/camelize/blob/main/index.js#L11) as well)

File: types/camelize/index.d.ts [hunk 5]
```diff
@@ -7,14 +7,14 @@ type CamelCase<S extends string> = S extends `${infer FirstWord}${CamelSeparator
...
+    : T extends object ? {
```
Comment (lukaskral) [inline types/camelize/index.d.ts:13]: I'm changing here `T extends {}` to `T extends object` which I believe is more accurate

**Review by tristan-f-r — APPROVED** [review]

**Comment by lukaskral** [comment]
Ready to merge

**Pull request merged**